stablehlo/transforms/StablehloRefineShapes.cpp

```diff
@@ -526,6 +526,28 @@ LogicalResult refineReturnShape(PatternRewriter& rewriter, OpType op,
   return refineReturnShape(rewriter, op, shape);
 }
 
+struct RefineAllGatherOpPattern : public OpRewritePattern<AllGatherOp> {
+  using OpRewritePattern::OpRewritePattern;
+  LogicalResult matchAndRewrite(AllGatherOp op,
+                                PatternRewriter& rewriter) const override {
+    auto operandType = op.getOperand().getType().cast<ShapedType>();
+    if (!operandType.hasRank())
+      return rewriter.notifyMatchFailure(op, "expected ranked operand type");
+
+    if (op.getChannelHandle() && !op.getUseGlobalDeviceIds())
```

Comment:
A comment like "This combination of attributes requires knowing `num_partitions` and cannot be optimized". Would be helpful in capturing the rationale of this change. I find myself flipping between spec and PR description several times to understand this code.

Reply:
Added a comment. Thank you for the feedback!